diff -up fuse-2.8.3/util/fusermount.c.regfix fuse-2.8.3/util/fusermount.c
--- fuse-2.8.3/util/fusermount.c.regfix	2011-08-29 09:43:14.929047002 -0400
+++ fuse-2.8.3/util/fusermount.c	2011-08-29 09:46:57.269977261 -0400
@@ -411,7 +411,7 @@ static int unmount_fuse_locked(const cha
 {
 	int currdir_fd = -1;
 	char *copy;
-	const char *last;
+	const char *last = mnt;
 	int umount_flags = lazy ? UMOUNT_DETACH : 0;
 	int res;
 
